File: combine_permits/combine_same_geolocation.m
 clear
 [~,~,raw]=xlsread('04apr2014_combined_geocoded.csv');
 
 %find columns
 iname = find(ismember(raw(1,:),'Center Name'));
 iholder = find(ismember(raw(1,:),'Permit Holder'));
 iaddress = find(ismember(raw(1,:),'Address'));
 izip= find(ismember(raw(1,:),'Zip Code'));
 iphone = find(ismember(raw(1,:),'Phone'));
 ipermit = find(ismember(raw(1,:),'Permit Number'));
 iexp = find(ismember(raw(1,:),'Permit Expiration Date'));
 istatus = find(ismember(raw(1,:),'Permit Status'));
 iagerange= find(ismember(raw(1,:),'Age Range'));
 icap = find(ismember(raw(1,:),'Maximum Capacity'));
 imed = find(ismember(raw(1,:),'Certified To Administer Medication'));
 itype = find(ismember(raw(1,:),'Site Type'));
 ilon = find(ismember(raw(1,:),'Lon'));
 ilat = find(ismember(raw(1,:),'Lat'));
 
 %make zip codes into strings
 raw(2:end,izip) = cellfun(@num2str,raw(2:end,izip),'UniformOutput',false);
 %make lat and lon into strings, easier to handle cell arrays of strings
 raw(2:end,ilon) = cellfun(@num2str,raw(2:end,ilon),'UniformOutput',false);
 raw(2:end,ilat) = cellfun(@num2str,raw(2:end,ilat),'UniformOutput',false);
 
 combdata = raw(1,:);
 k = 2;
 for i = 2: length(raw)
     if isempty(find(ismember(combdata(:,ilat),cell2mat(raw(i,ilat)))&ismember(combdata(:,ilon),cell2mat(raw(i,ilon)))))
        %isempty(find(ismember(combdata(:,iaddress),cell2mat(raw(i,iaddress)))&ismember(combdata(:,izip),cell2mat(raw(i,izip)))))
        %add entry if lat long is not the same as any other entry
        combdata(k,:) = raw(i,:);
%         %adjust format of expiration date
%         expdate = cell2mat(raw(i,iexp));
%         expdatenum = num2str(expdate/10^11);
%         year = expdatenum(1:4);
%         month = expdatenum(5:6);
%         combdata(k,iexp) = {[year,'-',month]};
        %make permit number a string
        combdata(k,ipermit) = {num2str(cell2mat(combdata(k,ipermit)))};
        k = k+1;
     else
        %locate other entry with same lat, lon
        i_comb = find(ismember(combdata(:,ilat),cell2mat(raw(i,ilat)))&ismember(combdata(:,ilon),cell2mat(raw(i,ilon))));
        %locate other entry with same address
        %problem is with closeby addresses but same lat lon
        %i_comb = find(ismember(combdata(:,iaddress),cell2mat(raw(i,iaddress)))&ismember(combdata(:,izip),cell2mat(raw(i,izip))));
        %combine age ranges of two permits
        combagerange = [raw(i,iagerange),combdata(i_comb,iagerange)];
        minage = min([str2num(combagerange{1}(1)),str2num(combagerange{2}(1))]);
        maxage = max([str2num(combagerange{1}(11)),str2num(combagerange{2}(11))]);
        combstr = cell2mat(combdata(i_comb,iagerange));
        combstr(1) = num2str(minage);
        combstr(11) = num2str(maxage);
        combdata(i_comb,iagerange) = {combstr};
        combdata(i_comb,icap) = {cell2mat(combdata(i_comb,icap))+cell2mat(raw(i,icap))};
%         %adjust format of expiration date
%         expdate = cell2mat(raw(i,iexp));
%         expdatenum = num2str(expdate/10^11);
%         year = expdatenum(1:4);
%         month = expdatenum(5:6);
        %combine expiration dates of both premits
        combdata(i_comb,iexp) = {[cell2mat(combdata(i_comb,iexp)),' / ',cell2mat(raw(i,iexp))]};
        %combine permit statuses of both permits
        combdata(i_comb,istatus) = {[cell2mat(combdata(i_comb,istatus)),' / ',cell2mat(raw(i,istatus))]};
        %combine permit number of both permits
        combdata(i_comb,ipermit) = {[num2str(cell2mat(combdata(i_comb,ipermit))),' / ',num2str(cell2mat(raw(i,ipermit)))]};
        if ~strcmp(combdata(i_comb,iname),raw(i,iname))
            %combine center names if not the same
            combdata(i_comb,iname) = {[cell2mat(combdata(i_comb,iname)),' / ',cell2mat(raw(i,iname))]};
        end
        if ~strcmp(combdata(i_comb,iholder),raw(i,iholder))
            %combine permit holders if not the same
            combdata(i_comb,iholder) = {[cell2mat(combdata(i_comb,iholder)),' / ',cell2mat(raw(i,iholder))]};
        end
        if ~strcmp(combdata(i_comb,iphone),raw(i,iphone))
            %combine phone number if not the same
            combdata(i_comb,iphone) = {[cell2mat(combdata(i_comb,iphone)),' / ',cell2mat(raw(i,iphone))]};
        end
        if ~strcmp(combdata(i_comb,iaddress),raw(i,iaddress))
            %combine address if not the same (despite same geolocation)
            combdata(i_comb,iaddress) = {[cell2mat(combdata(i_comb,iaddress)),' / ',cell2mat(raw(i,iaddress))]};
        end
        if ~strcmp(combdata(i_comb,itype),raw(i,itype))
            %combine permit types if not the same
            combdata(i_comb,itype) = {[cell2mat(combdata(i_comb,itype)),' / ',cell2mat(raw(i,itype))]};
        end
        if ~strcmp(combdata(i_comb,imed),raw(i,imed))
            %combine medication permit if not the same
            combdata(i_comb,imed) = {[cell2mat(combdata(i_comb,imed)),' / ',cell2mat(raw(i,imed))]};
        end
     end
 end
 
 xlswrite('04apr2014_combined_geocoded.xls',combdata);

File: python_combine_and_geocode/combine_same_geolocation.py
import sys
import csv

def combine_samegeo(filename):
    with open(filename, 'rb') as input_file:
        info = csv.reader(input_file)
        header = info.next()
        iname = header.index('Center Name')
        iholder = header.index('Permit Holder')
        iaddress = header.index('Address')
        izip= header.index('Zip Code')
        iphone = header.index('Phone')
        ipermit = header.index('Permit Number')
        iexp = header.index('Permit Expiration Date')
        istatus = header.index('Permit Status')
        iagerange= header.index('Age Range')
        icap = header.index('Maximum Capacity')
        imed = header.index('Certified to Administer Medication')
        itype = header.index('Site Type')
        ilat = header.index('Lat')
        ilon = header.index('Lon')
        itotcap = header.index('Total Maximum Capacity')

        newrows = []  # Initialize clean rows
        coord = []  # Initialize coordinate tupels
        # add entry only if geolocation hasn't been added yet
        for row in info:
            if (row[ilat], row[ilon]) not in coord:
                coord.append((row[ilat], row[ilon]))
                newrows.append(row)
            else:
                i_comb = coord.index((row[ilat], row[ilon]))
                combname = row[iname] + ' / ' + newrows[i_comb][iname]
                newrows[i_comb][iname] = combname
                combholder = row[iholder] + ' / ' + newrows[i_comb][iholder]
                newrows[i_comb][iholder] = combholder
                combaddress = row[iaddress] + ' / ' + newrows[i_comb][iaddress]
                newrows[i_comb][iaddress] = combaddress
                combphone = row[iphone] + ' / ' + newrows[i_comb][iphone]
                newrows[i_comb][iphone] = combphone
                combpermit = row[ipermit] + ' / ' + newrows[i_comb][ipermit]
                newrows[i_comb][ipermit] = combpermit
                combexp = row[iexp] + ' / ' + newrows[i_comb][iexp]
                newrows[i_comb][iexp] = combexp
                combstatus = row[istatus] + ' / ' + newrows[i_comb][istatus]
                newrows[i_comb][istatus] = combstatus
                combagerange = row[iagerange] + ' / ' + newrows[i_comb][iagerange]
                newrows[i_comb][iagerange] = combagerange
                combcap = row[icap] + ' / ' + newrows[i_comb][icap]
                newrows[i_comb][icap] = combcap
                combmed = row[imed] + ' / ' + newrows[i_comb][imed]
                newrows[i_comb][imed] = combmed
                combtype = row[itype] + ' / ' + newrows[i_comb][itype]
                newrows[i_comb][itype] = combtype
                newrows[i_comb][itotcap] = int(newrows[i_comb][itotcap]) + int(row[itotcap])

    nname = filename[:-4] + "_combinedgeo.csv" # The filename of the output file

    with open(nname, "wb") as output_file:
        writer = csv.writer(output_file)
        writer.writerow(header)
        writer.writerows(newrows)
    return nname

def main():
    fname = sys.argv[1]
    nname = combine_samegeo(fname)
    return nname
    
if __name__ == '__main__':
    main()